core/src/main/java/software/amazon/awssdk/SdkRequestOverrideConfig.java

```diff
@@ -0,0 +1,131 @@
+package software.amazon.awssdk;
+
+import software.amazon.awssdk.event.ProgressListener;
+import software.amazon.awssdk.utils.Validate;
+import software.amazon.awssdk.utils.builder.CopyableBuilder;
+import software.amazon.awssdk.utils.builder.ToCopyableBuilder;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+/**
+ * Base per-request override configuration for all SDK requests.
+ */
+public abstract class SdkRequestOverrideConfig<B extends SdkRequestOverrideConfig.Builder<B, C>,
+        C extends SdkRequestOverrideConfig<B, C>> implements ToCopyableBuilder<B, SdkRequestOverrideConfig<B, C>> {
+
+    private final ProgressListener progressListener;
+
+    private final Map<String, String> customHeaders;
+
+    private final Map<String, List<String>> customQueryParameters;
+
+    protected SdkRequestOverrideConfig(B builder) {
+        this.progressListener = builder.getProgressListener();
+        this.customHeaders = builder.getCustomHeaders();
+        this.customQueryParameters = builder.getCustomQueryParameters();
+    }
+
+    public Optional<ProgressListener> progressListener() {
+        return Optional.ofNullable(progressListener);
+    }
+
+    public Optional<Map<String, String>> customHeaders() {
+        return Optional.ofNullable(customHeaders);
+    }
```

Comment:
In the HTTP request/response we have the default be an empty Map<String, List<String>> instead of optional. Should we do that here too? (Same for query params)

Additional nitpick: maybe additionalHeaders and additionalQueryParameters? Just to make it clearer we're going to merge these in. We should make sure the name matches whatever is in the client configuration.

Reply:
+1 i'll make both changes.